### ISSUE TITLE
[Document] Properties for Area-Bricks

### ISSUE THREAD

**Issue by brusch** [pimcore/pimcore#6370]
Similar to properties on elements and the successor of the `edit.php`. 
The `edit.php` is sometimes quite hard to configure/setup (layout issues, CSS issues, ...), so it makes more sense to move this functionality to a more structured and configurable concept. 

This is how it should work: 

There should be a button to configure / edit the properties the area brick: 
![image](https://user-images.githubusercontent.com/142037/81550833-188a4300-9381-11ea-9709-906153a82d55.png)

To enable this button the area brick class has to implement a certain interface: 
```php
<?php

namespace AppBundle\Document\Areabrick;

use Pimcore\Extension\Document\Areabrick\PropertySupportInterface;

class Blockquote extends AbstractAreabrick implements PropertySupportInterface
{
    
}
```

Implementing this interface allows you to freely define properties (similar to the element properties). 
If you'd like to preconfigure properties you have to implement `PropertyConfigurationInterface`: 

```php
<?php

namespace AppBundle\Document\Areabrick;

use Pimcore\Extension\Document\Areabrick\PropertySupportInterface;
use Pimcore\Extension\Document\Areabrick\PropertyConfigurationInterface;
use Pimcore\Model\Document\Tag\Area\Info;

class Blockquote extends AbstractAreabrick implements PropertySupportInterface
{
    /**
     * @param Info $info
     * @param array $params contains the custom brick params defined in areablock option `params` or `globalParams`
     * @return array
     */
    public function configureProperties(Info $info, array $params = []):array
    {
        return [
            'myName' => [
                'type' => 'input',
                'description' => 'foo bar awsome ...', // optional
                'default' => 'my custom text'
            ],
            'mySelectProp' => [
                'type' => 'select',
                'options' => [
                    'key1' => 'value1',
                    'key2' => 'value2',
                ]
            ]
        ];
    }
}
```

This is how the grid should look like. 
The description should be displayed on hover. 

![image](https://user-images.githubusercontent.com/142037/81552664-2d1c0a80-9384-11ea-956a-992685311173.png)

Persisting: 
Maybe serializing them into the areablock is the best way to persist the values.

**Comment by solverat** [comment]
That's one of the main reasons why we‘ve invented the toolbox some years ago 👌.

Back to topic: How do you manage „relations“ within this grid view? I also prefer symfony configuration over php api - hopefully we‘ll find a good way to migrate to this new feature!

**Comment by dpfaffenbauer** [comment]
Copy paste Toolbox into Pimcore and done? :P

**Comment by brusch** [comment]
@solverat regarding Symfony config ... I think this would just add another layer of complexity if this would be the standard. Since the brick class is anyway a service, you can use the config anyway if you prefer to do so, right?

**Comment by dpfaffenbauer** [comment]
@brusch using Symfony Config makes the usage for the BrickClass obsolete for very simple bricks without any logic. which is quite often the case.

**Comment by brusch** [comment]
But the brick class is needed anyway, isn't it? 🤔

**Comment by dpfaffenbauer** [comment]
one is needed, but since bricks get registered using Symfony Services, you only need one for a very simple class like:

```yml
app.document.areabrick.teaser:
        class: AppBundle\Document\Areabrick\SimpleBrick
        parent: ToolboxBundle\Document\Areabrick\AbstractAreabrick
        calls:
            - [setName, ['tilo: Teaser']]
        tags:
            - { name: pimcore.area.brick, id: tilo_teaser }

    app.document.areabrick.custom_teaser:
        class: AppBundle\Document\Areabrick\SimpleBrick
        parent: ToolboxBundle\Document\Areabrick\AbstractAreabrick
        calls:
            - [setName, ['tilo: Custom Teaser']]
        tags:
            - { name: pimcore.area.brick, id: tilo_custom_teaser }

    app.document.areabrick.teaser_box:
        class: AppBundle\Document\Areabrick\SimpleBrick
        parent: ToolboxBundle\Document\Areabrick\AbstractAreabrick
        calls:
            - [setName, ['tilo: Teaser Box']]
        tags:
            - { name: pimcore.area.brick, id: tilo_teaser_box }

    app.document.areabrick.support:
        class: AppBundle\Document\Areabrick\SimpleBrick
        parent: ToolboxBundle\Document\Areabrick\AbstractAreabrick
        calls:
            - [setName, ['tilo: Support']]
        tags:
            - { name: pimcore.area.brick, id: tilo_support }
```

and the configuration with toolbox:

```yml
tilo_teaser:
    config_elements:
        headline:
            type: input
            title: 'Headline'
        content:
            type: textarea
            title: 'Content'
        backgroundImage:
            type: image
            title: 'Background Image'
        link:
            type: href
            title: 'Links'
            config:
                types: [document]
                subtypes:
                    document: ['page', 'link']
        colorType:
            type: select
            title: 'Color Type'
            config:
                store:
                    parkett: 'Parkett'
                    natur: 'Natur'
                    design: 'Design'
                    yellow: 'Yellow'
                default: 'yellow'
tilo_custom_teaser:
    config_elements:
        headline:
            type: input
            title: 'Headline'
        content:
            type: wysiwyg
            title: 'Content'
        links:
            type: multihref
            title: 'Links'
            config:
                types: [document]
                subtypes:
                    document: ['page', 'link']
        colorType:
            type: select
            title: 'Color Type'
            config:
                store:
                    parkett: 'Parkett'
                    natur: 'Natur'
                    design: 'Design'
                    yellow: 'Yellow'
                default: 'yellow'
tilo_teaser_box:
    config_elements:
        headline:
            type: input
            title: 'Headline'
        content:
            type: textarea
            title: 'Content'
        svg:
            type: href
            title: 'SVG'
        image:
            type: image
            title: 'Image'
        color:
            type: select
            title: 'Color'
            config:
                store:
                    natur: 'Green'
                    parkett: 'Red'
                    design: 'Blue'
                    yellow: 'Yellow'
                default: yellow
        link:
            type: href
            title: 'Link'
            config:
                types: [document]
                subtypes:
                    document: ['page', 'link']
        external_link:
            type: link
            title: 'External Link'
tilo_support:
    config_elements:
        backgroundText:
            type: input
            title: 'Background Text'
        headline:
            type: input
            title: 'Headline'
        headlineUnderlined:
            type: input
            title: 'Headline underlined'
        content:
            type: wysiwyg
            title: 'Content'
        image:
            type: image
            title: 'Image'
        content_extra:
            type: wysiwyg
            title: 'Content after Image'
```

so that saves me quite some time adding a new brick

**Comment by passioneight** [comment]
We suggest to add another `type` to the _Predefined Properties_, namely `brick`, as defining properties this way was already introduced and, thus, keeps configuration consistent. Additionally, if the need arises, one can easily re-use the configured brick-properties.

If some properties need to be added to the brick by default, one could implement @brusch's `configureProperties` like so:

```php
public function configureProperties(Info $info, array $params = []):array
{
     return [
         \Pimcore\Model\Property\Predefined::getByKey("property-key"),
         \Pimcore\Model\Property\Predefined::getByKey("another-property-key")
    ];
}
```

**Comment by brusch** [comment]
I'd go for the `configureProperties()` approach first, because it's much more flexible (e.g. for context sensitive properties, ...) than the static Symfony Config approach. It additionally allows the usage of the Symfony Config if someone prefers to do so. We can even discuss to additionally add this config feature later on, but for the start I'd just go for the first approach.

**Comment by msphn** [comment]
Just want to tell you that I really love that configureProperties approach, finally centralizing the configuration of an areabrick and reducing complexity in twig is a huge step forward. Thanks a lot! (Sorry for offtopic)

**Comment by dvesh3** [comment]
Resolved by #6923
